### PR TITLE
fix: listtransactions crash #3231

### PR DESCRIPTION
**Pull request by m-schmoock**
Fixes https://github.com/ElementsProject/lightning/issues/3231

Ready for Review and Merge

### CODE REVIEW

**Comment by m-schmoock** [comment]
Note: 1c5c305 just gets rid of the warning `**BROKEN** lightningd(2455): Accessing a null column ...`

There is still a more or less reproducible memory access/corruption. Current test does not reflect this...

File: wallet/wallet.c [hunk 1]
```diff
@@ -3445,26 +3445,27 @@ struct wallet_transaction *wallet_transactions_get(struct wallet *w, const tal_t
...
-			cur = &txs[count];
+			tal_resize(&txs, tal_count(txs) + 1);
+			cur = &txs[tal_count(txs) - 1];
 			db_column_txid(stmt, 0, &cur->id);
 			cur->tx = db_column_tx(txs, stmt, 1);
 			cur->rawtx = tal_dup_arr(txs, u8, db_column_blob(stmt, 1),
 						 db_column_bytes(stmt, 1), 0);
-			cur->blockheight = db_column_int(stmt, 2);
-			cur->txindex = db_column_int(stmt, 3);
+            if (!db_column_is_null(stmt, 2))    // MSC: tx may be unconfirmed
```
Comment (cdecker) [inline wallet/wallet.c]:
```suggestion
            if (!db_column_is_null(stmt, 2)) /* TX may be unconfirmed. */
```

File: wallet/wallet.c [hunk 2]
```diff
@@ -3445,26 +3445,27 @@ struct wallet_transaction *wallet_transactions_get(struct wallet *w, const tal_t
...
+                cur->blockheight = db_column_int(stmt, 2);
+            if (!db_column_is_null(stmt, 3))
+                cur->txindex = db_column_int(stmt, 3);
```
Comment (cdecker) [inline wallet/wallet.c]: The two `if` statements should likely be combined (and an assert added): we should never have a `blockheight` without also having a `txindex`, and vice-versa.

Comment (m-schmoock) [reply wallet/wallet.c]: ok

File: wallet/wallet.c (hunk 1 above)
Comment (m-schmoock) [reply wallet/wallet.c]: ok

File: CHANGELOG.md [hunk 3]
```diff
@@ -79,6 +79,7 @@ Note: You should always set `allow-deprecated-apis=false` to test for changes.
 - Fixed bogus "Bad commit_sig signature" which caused channel closures when reconnecting after updating fees under simultaneous bidirectional traffic.
 - Relative `--lightning_dir` is now working again.
 - Build: MacOS now builds again (missing pwritev).
+- Fixed `listtransactions` crash for unconfirmed channels or transactions.
```
Comment (cdecker) [inline CHANGELOG.md]: We recently started putting changelog entries in the commit messages that implement the changes (https://github.com/ElementsProject/lightning/pull/3238) this way we avoid the perennial conflicts on the `CHANGELOG.md` file :wink:

Comment (m-schmoock) [reply CHANGELOG.md]: thanks for the update :D

**Comment by m-schmoock** [comment]
okay, done from my side...

**Comment by cdecker** [comment]
Awesome, thanks @m-schmoock :+1:

**Comment by m-schmoock** [comment]
Yes, but it can now raises consistency assertions on 'older' databases:

https://github.com/ElementsProject/lightning/issues/3265

**Comment by m-schmoock** [comment]
@cdecker Yes, but it can now raise consistency assertions on 'older' databases:

https://github.com/ElementsProject/lightning/issues/3265

Ideas? I can fix that too :D ...

**Pull request merged**